### ISSUE TITLE
[FEATURE] TPM support

### ISSUE THREAD

**Issue by abonillabeeche** [harvester/harvester#2910]
TPM Support is expected for certain Edge/secure deployments (FIDO Secure Onboarding).
Windows 11 requires TPM
Elemental requires TPM

**Comment by abonillabeeche** [comment]
This is achievable via virt-manager and this is an example with screenshots
https://rancher.github.io/elemental/tpm/#add-tpm-emulation-to-bare-metal-machine

**Comment by futuretea** [comment]
Refer to https://kubevirt.io/2022/KubeVirt-installing_Microsoft_Windows_11_from_an_iso.html
We need an option to add emulated  TPM to VM in both UI and terraform-provider-harvester

**Comment by Martin-Weiss** [comment]
How can we add tpm: {} in the device section of a VM using the terraform provider harvester_virtualmachine (Resource) for harvester?

**Comment by futuretea** [comment]
> How can we add tpm: {} in the device section of a VM using the terraform provider harvester_virtualmachine (Resource) for harvester?

I will submmit a PR to support it.

**Comment by abonillabeeche** [comment]
@futuretea did you manage to get this PR submitted? Wondering if will add `tpm: {}` to both a single instance and multiple instances when creating VMs.

**Comment by abonillabeeche** [comment]
@rebeccazzzz can we move this to 1.1.2 considering the low effort?

**Comment by irishgordo** [comment]
**Validation Failed**
@futuretea @WuJun2016 

Currently, trying to validate this on Harvester Version: master-c9d8825c-head following [Windows Install Instructions](https://docs.harvesterhci.io/dev/vm/create-windows-vm)

And running into an issue where Windows 11 Enterprise Edition, based on [Windows 11 Enterprise Evaluation, available from Microsoft](https://www.microsoft.com/en-us/evalcenter/evaluate-windows-11-enterprise) and I'm not able to complete the install with TPM enabled.
Also seeing issues enabling EFI & Secure Boot, w/ TPM Enabled.

![Screenshot from 2023-05-15 15-13-14](https://github.com/harvester/harvester/assets/5370752/5e1de356-df22-4bd2-bf53-f6a0f91db635)

**Comment by futuretea** [comment]
@irishgordo Can you help to confirm whether the VM specifications used meet [windows-11-requirements](https://learn.microsoft.com/en-us/windows/whats-new/windows-11-requirements#hardware-requirements) ?

**Comment by futuretea** [comment]
@irishgordo 
Refer to https://github.com/harvester/docs/pull/310
Both the TPM device and UEFI firmware with SecureBoot are hard requirements for Windows 11.
![2953e1f05f55d4ada6f5481eef90722](https://github.com/harvester/harvester/assets/15064560/56f4682d-3f3b-484b-9354-0d485eedb637)
![92223a1dcc37ce124625182e7944849](https://github.com/harvester/harvester/assets/15064560/f9cf8d00-e840-49b1-8dc1-2f9d1299bd8f)

**Comment by irishgordo** [comment]
**validation in progress, slight issues, will check with latest master-head**

I'm noticing issues migrating but will be retrying the migration with the latest master-head release as:
https://github.com/harvester/harvester/pull/3399

Just got merged in earlier :smile: 

Also noticing that a WindowVM created -> create a snapshot -> restore to a new vm from snapshot 
The vm will never spin up.
And will hang.
I will check if I can reproduce on the latest master-head prior to opening an issue :smile: 

![Screenshot from 2023-05-16 12-30-09](https://github.com/harvester/harvester/assets/5370752/9df68bc7-d075-4ff1-ba1c-32ba87b59d40)
![Screenshot from 2023-05-16 12-29-08](https://github.com/harvester/harvester/assets/5370752/00258e38-facd-42aa-bd94-ffbb6c79a798)
![Screenshot from 2023-05-16 12-28-04](https://github.com/harvester/harvester/assets/5370752/b36c7bab-ec7c-46ae-9348-5bed39d41049)
![Screenshot from 2023-05-16 12-17-16](https://github.com/harvester/harvester/assets/5370752/fcd16e23-cdc7-45d2-9bbf-66d3449470d2)
![Screenshot from 2023-05-16 12-16-31](https://github.com/harvester/harvester/assets/5370752/5038a62d-5f3f-4e27-80e7-19b420a4d3c5)
![Screenshot from 2023-05-16 12-14-06](https://github.com/harvester/harvester/assets/5370752/5093fe3c-0338-40a5-9fdb-43029dbf70a2)
![Screenshot from 2023-05-16 12-07-37](https://github.com/harvester/harvester/assets/5370752/120b56e5-0d79-4c25-9746-b494d269b2b1)
![Screenshot from 2023-05-16 11-54-14](https://github.com/harvester/harvester/assets/5370752/2d873598-a2f7-4fea-a914-1528208f20e2)
[supportbundle_d0a70753-473a-4de6-8b8c-24649bded98d_2023-05-16T19-42-24Z.zip](https://github.com/harvester/harvester/files/11491659/supportbundle_d0a70753-473a-4de6-8b8c-24649bded98d_2023-05-16T19-42-24Z.zip)

**Comment by irishgordo** [comment]
**Validation Failed**

@futuretea it seems that migration between nodes does not happen from the following scenario:
- create windows vm
- install drivers for virtio
- install windows
- allow to boot
- eject both CD-ROM containing Windows 11 ISO & Container Disk `registry.suse.com/suse/vmdp/vmdp:2.5.3`
- attempt to migrate VM

![Screenshot from 2023-05-16 16-10-31](https://github.com/harvester/harvester/assets/5370752/1e0b1e64-a652-45f8-b2ac-8d7e5f31053c)
![Screenshot from 2023-05-16 16-09-56](https://github.com/harvester/harvester/assets/5370752/1d9595a0-96a3-435a-b0d7-645fc05d8676)

```
curl 'https://HARVESTERVIPIP/v1/harvester/kubevirt.io.virtualmachines/default/win-11-ent-eval-test?action=migrate' -X POST -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/112.0' -H 'Accept: application/json' -H 'Accept-Language: en-US,en;q=0.5' -H 'Accept-Encoding: gzip, deflate, br' -H 'Content-Type: application/json;charset=utf-8' -H 'x-api-csrf: d3511a4b1642481199ce7161e5fded4c' -H 'Origin: https://HARVESTERVIPIP' -H 'Connection: keep-alive' -H 'Referer: https://HARVESTERVIPIP/dashboard/harvester/c/local/kubevirt.io.virtualmachine' -H 'Cookie: R_PCS=light; R_LOCALE=en-us; R_REDIRECTED=true; CSRF=d3511a4b1642481199ce7161e5fded4c; R_SESS=token-44zqf:cdrgftlk92sn7nzvf859s5gfthtvph6645x5scrqjn4w56thfwfvss' -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: same-origin' -H 'TE: trailers' --data-raw '{"nodeName":"harvester-hzb6c"}'
```
Yields `500`:
```
Volume disk-2 is container disk, needs to be removed before migration
```

There seems to be no, disk-2 attached to the VM currently however.

Tested with Harvester  Version: master-0b495171-head.
That should have:
https://github.com/harvester/harvester/pull/3399
Within the source I believe :thinking:

**Comment by futuretea** [comment]
@irishgordo View yaml by editing yaml, found container disk, It's a bug about `enjectCDROM`, not introduced by https://github.com/harvester/harvester/pull/3399, It has been around for a long time, but no one ever found it. Good job! 
![image](https://github.com/harvester/harvester/assets/15064560/14976b6a-58de-4f70-81ed-116d638ddf30)

**Comment by futuretea** [comment]
@irishgordo issue https://github.com/harvester/harvester/issues/3914 created, ~I will fix it~, fixed

**Comment by irishgordo** [comment]
![Screenshot from 2023-05-17 12-24-01](https://github.com/harvester/harvester/assets/5370752/31080878-043c-4ae8-b847-cd2575fe7512)
![Screenshot from 2023-05-17 12-23-24](https://github.com/harvester/harvester/assets/5370752/515043e6-2ee7-4e87-802a-5e193291bb4d)
![Screenshot from 2023-05-17 12-22-26](https://github.com/harvester/harvester/assets/5370752/463c58c2-1141-4164-b0de-b46e4db4c0d8)
![Screenshot from 2023-05-17 12-20-42](https://github.com/harvester/harvester/assets/5370752/bb6aa8e9-603d-4c44-9d8d-f760f2159148)
![Screenshot from 2023-05-17 12-20-25](https://github.com/harvester/harvester/assets/5370752/18f1c4f8-6c86-4763-8b69-fe854e3e7e83)
![Screenshot from 2023-05-17 12-20-09](https://github.com/harvester/harvester/assets/5370752/83d966ff-b294-4ca3-8185-e9bf6e0454d9)

@futuretea thanks for the fix :+1: :smile:  in #3914 

Tested that migration now works off of: Harvester Version: master-25d478c0-head 

**Validation Failed**

But I just opened:
https://github.com/harvester/harvester/issues/3929

As noticing (Win11 TPM VM) VM A -> Snapshot of A -> Restore to New VM (B) -> Take Snapshot of New VM (B)  -> Stop New VM (B) -> Replace Existing (B)
Is hanging/failing

**Comment by irishgordo** [comment]
@futuretea thanks for the fix surrounding #3929 - I was able to validate on master-head: master-d739c2a4-head that the problem no longer exists and everything else surrounding this looks good :+1: :smile: 

![Screenshot from 2023-05-18 16-50-57](https://github.com/harvester/harvester/assets/5370752/800158e6-411b-40e8-95bb-9f0d6135ec21)
![Screenshot from 2023-05-18 15-50-46](https://github.com/harvester/harvester/assets/5370752/d8c25c31-b69f-4c18-ac44-113019f3fd39)

This looks good from my end :smile:  - I'll go ahead and close this out
